### ISSUE TITLE
Unstable VRT for buttons 

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#3552]
## Bug Description

The visual regression test for buttons has become somewhat unstable recently and fails intermittently. Specifically, the failure is related to the button in a hovered state.

## Screenshots

![image](https://user-images.githubusercontent.com/1621608/121512176-1f570f80-c9f2-11eb-8165-5ba693785d6b.png)

_Backstop reference image diff with test_

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* Visual regression tests for buttons should consistently pass

### Test Coverage

* 

### Visual Regression Changes

* Run visual regression tests and check that the buttons now pass reliably

## QA Brief

* 

## Changelog entry

*

**Comment by aaemnnosttv** [comment]
@Hazlitte we should not change the threshold for error to fix this, since that is essentially ignoring the problem and would reduce the effectiveness of all VRT tests. The same VRT test passes most of the time, so this issue is about figuring out why that happens and fixing this one case.

**Comment by Hazlitte** [comment]
@aaemnnosttv I have spent some more time investigating this and I think that the problem is related to `hoverSelector` not working for the buttons stories using Storybook 6.  I have tried passing the necessary parameters in a number of different ways.  I think @asvinb said that he has been solving implementation problems with storybook 6, perhaps he knows a way to get this to work.

**Comment by johnPhillips** [comment]
I've thrown an hour at this and not got anywhere either, unfortunately. 
For what it's worth, I reached out to @asvinb in case he had any insights but he believes this is unrelated to the issue he is tackling with storybook.
All I can see is that we are using backstop's `hoverSelector` parameter to get at the hover state of the button and that it seems to be working inconsistently for this test (there are other tests that use it that I haven't seen fail in the same way).

**Comment by techanvil** [comment]
I've spent some time looking at this but have not managed to figure it out. I'm putting it back down again for now as it's starting to feels like a bit of a rabbit hole for time.

Ground covered:

**Speed up debug feedback loop:** I couldn't reproduce this locally, so was restricted to running the VRT tests via GitHub Actions to debug them. In order to not interfere with regular development I forked the repo to my own GitHub account. I then made some changes to get the workflow execution time down from ~6-8 mins to ~40-90 seconds (for a single run). After a while I realised it would be useful to get Backstop to just run in a loop until it failed, so did that too: 

- Forked repo on GH to my own account.
- Changed default branch of repo order to tweak workflows on a new branch.
- Removed unused workflows (everything in `.github/workflows` apart from `visual-regression.yml`).
- Removed `dist` & `node_modules` from `.gitignore`. Added `node_modules/puppeteer`.
- Built storybook and added `dist` to repo.
- Added `node_modules` to repo (except `puppeteer` as Chromium too large a file for GH).
- Tweaked `visual-regression.yml` to run `npm i` (to install puppeteer), `npm run backstopjs -- --storybook-host=./dist/` instead of `npm run test:visualtest`, in order to avoid building the Storybook every time.
- Run the tests in a loop until failure: `while npm run backstopjs -- --storybook-host=./dist/ ; do echo SUCCESS ; sleep 1 ; done ; exit 1`.

**Debugging:** I tried the following tweaks/mods - unfortunately, the tests continued to success/fail randomly as per usual:

`package.json`:
- Bumped to `backstopjs v5.4.4` (Chromium r901912).

`tests/backstop/config.js`:
- Checked Docker container runtime - verified ~7GB.
  - `dockerCommandTemplate: 'docker run --rm ubuntu free -m`
- Checked Docker `shm-size`, 64MB, tried bumping to 2GB - https://github.com/garris/BackstopJS#issues-with-chrome-headless-in-docker.
  - `dockerCommandTemplate: 'docker run --rm ubuntu df -h` and check the `shm` line.
  - Then include `--shm-size=2G` in the real `dockerCommandTemplate` value:
  - ```dockerCommandTemplate: `docker run --shm-size=2G --rm -i${``` ...
- Reduced `asyncCaptureLimit` and `asyncCompareLimit` to 1 to run the tests synchronously.

`tests/backstop/viewports.js`:

- Reduced to a single viewport (commented out the medium & large viewports).

`tests/backstop/engine_scripts/puppet/clickAndHoverHelper.js`:
- Replaced deprecated `waitFor` with `waitForSelector/waitForTimeout` - https://github.com/puppeteer/puppeteer/issues/6214.
- Tried adding extra `waitForTimeout( 1000 )` after `page.waitForSelector( hoverSelectorIndex )`.
- Add `try/catch` around each `await` to check for thrown errors - no error thrown on test failure.
- Tried dumping the frame tree before all & after each await to see if there was any diff between success & failure cases but there wasn't - https://github.com/puppeteer/puppeteer/blob/main/docs/api.md#class-frame.
- Tried waiting for the button with hover state after the call to `page.hover`:
  - `page.waitForSelector('.googlesitekit-button--hover.mdc-button--raised:hover')`
- Dumped button opacity and hover state after each `await`, resulting opacity 0 for non-hover and 0.08 for hover. Regardless of success/failure the DOM was reporting the hover opacity to have changed after the call to `page.hover` - the opacity change is what makes the visual difference for the button.
```
async function dumpOpacity( page, msg ) {
  const { opacity, isHover } = await page.evaluate(() => {
    const opacity = window.getComputedStyle( document.querySelector('.googlesitekit-button--hover'), ':before' ).opacity
    const isHover = !! document.querySelector('.googlesitekit-button--hover:hover');
    return { opacity, isHover };
  });
  console.log('OPACITY, ISHOVER', msg, opacity, isHover);
}
```

**Comment by aaemnnosttv** [comment]
Nice detective work @techanvil. This will be useful in next steps, thanks for the thorough write up 👍

**Comment by eugene-manuilov** [comment]
This ticket won't be relevant anymore once #4184 is merged. That PR replaces the hover state with the icon button. So, once it's merged, we can close this ticket.

**Comment by techanvil** [comment]
@eugene-manuilov that sounds good, for now at least. Let's hope we don't run into the same issue again :crossed_fingers:

**Comment by tofumatt** [comment]
With #4184 merged, this issue is no longer relevant and can be closed. 🙂
